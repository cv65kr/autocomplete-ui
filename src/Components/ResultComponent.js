import {h} from "preact"
import {actions} from "../actions"
import { connect } from 'unistore/preact'
import Hogan from 'hogan.js'
import groupBy from 'lodash/groupBy'

const defaultHtmlAttributes = {
    id: "apisearch-listbox",
    className: "as-result",
    role: "listbox",
    tabIndex: "-1"
};

/**
 * Suggested Search Component
 */
export const ResultComponent = connect('resultBoxOpen, items')(
    ({
         /** component props */
         datasets,

         /** store props */
         items,
         resultBoxOpen
    }) => {
        if (false === resultBoxOpen) {
            return <div
                {...defaultHtmlAttributes}
                style={{display: 'none'}}
            />;
        }

        let filteredItemsByType = groupBy(items, 'uuid.type');
        let filteredDatasetsByType = groupBy(datasets, 'type');

        return (
            <div
                {...defaultHtmlAttributes}
            >
                {Object.keys(filteredItemsByType).map(type =>
                    <div
                        className={`as-result__dataset as-result__dataset--${type}`}
                    >
                        {(filteredDatasetsByType[type][0].template.header)
                            ? <div
                                className="as-result__datasetHeader"
                                dangerouslySetInnerHTML={
                                    renderTemplate(
                                        filteredDatasetsByType[type][0].template.header,
                                        null
                                    )
                                }
                            />
                            : null
                        }

                        <ul
                            className="as-result__datasetItemsList"
                        >
                            {filteredItemsByType[type].map(item =>
                                <li
                                    className="as-result__datasetItem"
                                    tabIndex={-1}
                                    dangerouslySetInnerHTML={
                                        renderTemplate(
                                            filteredDatasetsByType[type][0].template.item,
                                            item
                                        )
                                    }
                                />
                            )}
                        </ul>
                    </div>
                )}
            </div>
        )
    }
);

/**
 * Compile template
 * @param template
 * @param data
 */
const renderTemplate = (template, data) => {
    let compiledTemplate = Hogan.compile(template);

    return {
        __html: compiledTemplate.render(data)
    }
};