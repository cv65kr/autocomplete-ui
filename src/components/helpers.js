import {h} from "preact";
import {groupBy, sortBy} from "lodash";
import Hogan from "hogan.js";


/**
 * Compile template
 * @param template
 * @param data
 */
export const renderTemplate = (template, data) => {
    let compiledTemplate = Hogan.compile(template);

    return {
        __html: compiledTemplate.render(data)
    }
};

/**
 * Set active index
 *
 * @param items
 * @param currentCursorIndex
 * @returns {{}}
 */
export const createDatasetByItemTypeAndSetActiveIndex = (
    items,
    currentCursorIndex
) => {
    let sortedItemsByType = sortBy(items, 'uuid.type');
    let itemsWithActiveIndex = sortedItemsByType.map((item, index) =>
        ({
            ...item,
            navigationIndex: (index + 1),
            isActive: ((index + 1) === currentCursorIndex)
        })
    );

    return groupBy(itemsWithActiveIndex, 'uuid.type');
};

/**
 * Build the poweredBy icon
 * It can be a boolean:
 *   if (false) do now show a poweredBy
 *   if (true) assign the default poweredBy
 *   else render a custom poweredBy template string
 *
 * @param poweredBy
 * @returns {XML}
 */
export const buildPoweredBy = (poweredBy = false) => {
    if (false === poweredBy) return null;

    if (true === poweredBy) {
        poweredBy = `<a href="http://apisearch.io">
            <img 
                src="http://apisearch.io/assets/media/powered-by.svg" 
                alt="Powered by Apisearch" 
                width="100px"
            />
        </a>`;
    }

    return <div
        className="as-poweredBy"
        dangerouslySetInnerHTML={
            renderTemplate(poweredBy, null)
        }
    />
};