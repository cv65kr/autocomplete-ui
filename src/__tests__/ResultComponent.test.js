import {h} from 'preact';
import {shallow} from 'preact-render-spy'
import {ResultComponent} from '../components/ResultComponent';
import {resultWithItems} from "./__mocks__/result.mock";

const datasets = [
    {
        type: 'mock',
        template: {
            header: '<h3>Mock Dataset</h3>',
            item: '<p>{{metadata.name}}</p>'
        }
    }
];

describe('<ResultComponent />', () => {
    let tree;

    beforeEach(() => {
        tree = shallow(
            <ResultComponent
                datasets={datasets}
                resultBoxOpen={false}
                poweredBy={false}
            />
        );
    });

    it('should render initial state', () => {
        expect(tree).toMatchSnapshot();
    });

    it('should render results and navigate', () => {
        tree.render(<ResultComponent
            datasets={datasets}
            poweredBy={true}
            {...resultWithItems}
        />);
        expect(tree).toMatchSnapshot();

        tree.render(<ResultComponent
            datasets={datasets}
            poweredBy={true}
            {...resultWithItems}
            currentCursorIndex={2}
        />);
        expect(tree).toMatchSnapshot();
    });
});