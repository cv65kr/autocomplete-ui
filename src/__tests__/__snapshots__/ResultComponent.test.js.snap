// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultComponent /> should render initial state 1`] = `
preact-render-spy (1 nodes)
-------
<div
  id="apisearch-listbox"
  class="as-result"
  role="listbox"
  tabIndex="-1"
  style="display: none;"
>
</div>

`;

exports[`<ResultComponent /> should render results and navigate 1`] = `
preact-render-spy (1 nodes)
-------
<div
  id="apisearch-listbox"
  class="as-result"
  role="listbox"
  tabIndex="-1"
>
  <div class="as-result__dataset as-result__dataset--mock">
    <div class="as-result__datasetHeader">
      <h3>Mock Dataset</h3>
    </div>
    <ul class="as-result__datasetItemsList">
      <li
        data-index={1}
        class="as-result__datasetItem"
      >
        <p>This is a mocked item</p>
      </li>
      <li
        data-index={2}
        class="as-result__datasetItem"
      >
        <p>This is a mocked item 2</p>
      </li>
      <li
        data-index={3}
        class="as-result__datasetItem"
      >
        <p>This is a mocked item 3</p>
      </li>
    </ul>
  </div>
  <div class="as-poweredBy">
    <a href="http://apisearch.io">
                <img 
                    src="http://apisearch.io/assets/media/powered-by.svg" 
                    alt="Powered by Apisearch" 
                    width="100px"
                />
            </a>
  </div>
</div>

`;

exports[`<ResultComponent /> should render results and navigate 2`] = `
preact-render-spy (1 nodes)
-------
<div
  id="apisearch-listbox"
  class="as-result"
  role="listbox"
  tabIndex="-1"
>
  <div class="as-result__dataset as-result__dataset--mock">
    <div class="as-result__datasetHeader">
      <h3>Mock Dataset</h3>
    </div>
    <ul class="as-result__datasetItemsList">
      <li
        data-index={1}
        class="as-result__datasetItem"
      >
        <p>This is a mocked item</p>
      </li>
      <li
        data-index={2}
        class="as-result__datasetItem as-result__datasetItem--active"
      >
        <p>This is a mocked item 2</p>
      </li>
      <li
        data-index={3}
        class="as-result__datasetItem"
      >
        <p>This is a mocked item 3</p>
      </li>
    </ul>
  </div>
  <div class="as-poweredBy">
    <a href="http://apisearch.io">
                <img 
                    src="http://apisearch.io/assets/media/powered-by.svg" 
                    alt="Powered by Apisearch" 
                    width="100px"
                />
            </a>
  </div>
</div>

`;
